flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/reuse/ScanReuser.java

```diff
@@ -165,7 +166,7 @@ public List<RelNode> reuseDuplicatedScan(List<RelNode> relNodes) {
             }
 
             int[][] allProjectFields = allProjectFieldSet.toArray(new int[0][]);
-            List<String> allMetaKeys = new ArrayList<>(allMetaKeySet);
+            List<String> allMetaKeys = sortMetadataKeys(allMetaKeySet, pickTable.tableSource());
```

Comment:
NIT: method name could be `enforceMetadataKeyOrder` or `copyMetadataKeyOrder`; sort implies there is a sort order like ascending.